src/PSRule.Rules.Azure/rules/Azure.VNET.Rule.ps1

```diff
@@ -107,4 +107,37 @@ Rule 'Azure.VNET.SubnetName' -Ref 'AZR-000267' -Type 'Microsoft.Network/virtualN
     }
 }
 
+# Synopsis: VNETs with a GatewaySubnet should have an AzureBastionSubnet to allow for out of band remote access to VMs.
+Rule 'Azure.VNET.BastionSubnet' -Ref 'AZR-000314' -Type 'Microsoft.Network/virtualNetworks' -If { IsGatewaySubnet } -Tag @{ release = 'GA'; ruleSet = '2022_12' } {
+    $subnets = @(GetVirtualNetworkSubnets)
+    if ($subnets | Where-Object { $_.name -eq 'AzureBastionSubnet' }) { $Assert.Pass() } else { $Assert.Fail().Reason($LocalizedData.BastionSubnetNotFound, $TargetObject.name) }
+}
+
 #endregion Virtual Network
+
+#region Helper functions
+
+function global:IsGatewaySubnet {
+    [CmdletBinding()]
+    [OutputType([System.Boolean])]
+    param ()
+    process {
+        ($TargetObject.Properties.subnets | Where-Object { $_.name -eq 'GatewaySubnet' }) -or
+        (@(GetSubResources -ResourceType 'Microsoft.Network/virtualNetworks/subnets' |
+            Where-Object { $_.name -eq 'GatewaySubnet' }))
+    }
+}
+
+function global:GetVirtualNetworkSubnets {
+    [CmdletBinding()]
+    [OutputType([PSObject])]
+    param ()
+    process {
+        $TargetObject.Properties.subnets
+        @(GetSubResources -ResourceType 'Microsoft.Network/virtualNetworks/subnets' | ForEach-Object {
+                [PSCustomObject]@{ name = $_.name }
+            })
```

Comment:
This should be enough to get subnet names as strings.

```suggestion
        $TargetObject.Properties.subnets | ForEach-Object { $_.name };
        GetSubResources -ResourceType 'Microsoft.Network/virtualNetworks/subnets' | ForEach-Object {
            $_.name
        }
```

Reply:
Overengineered from my side, lets do it this way instead.